File: 0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        # Step 1: Find the length of the linked list
        length = 0
        current = head
        while current:
            length += 1
            current = current.next
        
        # Step 2: Determine the base size of each part and the number of extra nodes
        part_size = length // k
        extra_nodes = length % k
        
        result = []
        current = head
        
        # Step 3: Split the list into parts
        for i in range(k):
            part_head = current
            # Determine the size of the current part
            current_part_size = part_size + (1 if i < extra_nodes else 0)
            
            # Advance the current pointer to the end of the current part
            for j in range(current_part_size - 1):
                if current:
                    current = current.next
            
            # If there is a node, we need to break the connection to the rest of the list
            if current:
                next_part = current.next
                current.next = None
                current = next_part
            
            # Append the current part's head to the result
            result.append(part_head)
        
        # Step 4: If k > length of the list, append None for remaining parts
        while len(result) < k:
            result.append(None)
        
        return result

File: Daily Questions/0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
// Helper function to create a new node
struct ListNode* createNode(int val) {
    struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));
    newNode->val = val;
    newNode->next = NULL;
    return newNode;
}

struct ListNode** splitListToParts(struct ListNode* head, int k, int* returnSize) {
    // Step 1: Calculate the length of the linked list
    int length = 0;
    struct ListNode* current = head;
    while (current != NULL) {
        length++;
        current = current->next;
    }
    
    // Step 2: Calculate the size of each part
    int baseSize = length / k;
    int extraNodes = length % k;

    // Step 3: Create an array to hold the result
    struct ListNode** result = (struct ListNode**)malloc(k * sizeof(struct ListNode*));
    
    current = head;
    
    // Step 4: Split the list into parts
    for (int i = 0; i < k; i++) {
        result[i] = current;
        int currentPartSize = baseSize + (i < extraNodes ? 1 : 0);
        
        // Move current to the end of the current part
        for (int j = 0; j < currentPartSize - 1; j++) {
            if (current != NULL) {
                current = current->next;
            }
        }
        
        // Break the link to the next part
        if (current != NULL) {
            struct ListNode* nextPart = current->next;
            current->next = NULL;
            current = nextPart;
        }
    }
    
    *returnSize = k;
    return result;
}